File: csharp/TheatricalPlayersRefactoringKata/StatementPrinter.cs
using LanguageExt;
using System;
using System.Collections.Generic;
using System.Globalization;
using static LanguageExt.Prelude;

namespace TheatricalPlayersRefactoringKata
{
    public class StatementPrinter
    {
        public string Print(
            Invoice invoice, 
            Dictionary<string, Play> plays)
        {
            return
                TextFormatter.Format(
                invoice.Performances
                    .Map(p => CreateStatement(p, plays, TextFormatter.ToTextFormat))
                    .Reduce((agg, line) => agg.Append(line)), invoice.Customer);
        }

        private static Statement CreateStatement(
            Performance perf,
            Dictionary<string, Play> plays,
            Func<Performance, Play, int, string> lineFormatter)
        {
            var play = plays[perf.PlayID];
            var amount = CalculateAmountFor(perf, play)
                    .Match(a => a, fail => throw new Exception("unknown type: " + play.Type));

            return new Statement(
                amount,
                CalculateCredits(perf, play),
                lineFormatter(perf, play, amount));
        }

        private static int CalculateCredits(Performance perf, Play play) =>
            CalculateDefaultCredits(perf) + ("comedy" == play.Type ? (int)Math.Floor((decimal)perf.Audience / 5) : 0);

        private static int CalculateDefaultCredits(Performance perf) => Math.Max(perf.Audience - 30, 0);

        private static Map<string, Func<Performance, int>> calculateMap = Map<string, Func<Performance, int>>(("tragedy", CalculateTragedyAmount), ("comedy", CalculateComedyAmount));

        private static Try<int> CalculateAmountFor(Performance perf, Play play) => () => calculateMap[play.Type](perf);

        private static int CalculateComedyAmount(Performance performance)
        {
            int amount = 30000;
            if (performance.Audience > 20)
            {
                amount += 10000 + 500 * (performance.Audience - 20);
            }
            amount += 300 * performance.Audience;

            return amount;
        }

        private static int CalculateTragedyAmount(Performance performance)
        {
            int amount = 40000;
            if (performance.Audience > 30)
            {
                amount += 1000 * (performance.Audience - 30);
            }

            return amount;
        }
    }
}

File: fsharp/TheatricalPlayersRefactoringKata/StatementPrinter.fs
namespace TheatricalPlayersRefactoringKata

open System
open System.Globalization

module StatementPrinter =
    type Performance = { 
        PlayId: string
        Audience: int
    }

    type Play = {
        Name: string
        Type: string
    }

    type Invoice = {
        Customer: string
        Performances: List<Performance>
    }

    type Statement = {
        Text: string
        Amount: int
        Credits: int
    }

    let formatProvider = CultureInfo.GetCultureInfo("en-US")

    let formatLineToText name amount audience = 
        String.Format(formatProvider, "  {0}: {1:C} ({2} seats)\n", name, amount / 100, audience)

    let formatStatement statement customer =        
        String.Format(formatProvider, "Statement for {0}\n{1}Amount owed is {2:C}\nYou earned {3} credits\n", customer, statement.Text, statement.Amount / 100, statement.Credits)

    let calculateComedyPrice audience =  
        30000 + (300 * audience) + if audience > 20 then 10000 + 500 * (audience - 20) else 0

    let calculateTragedyPrice audience =  
        if audience > 30 then 40000 + 1000 * (audience - 30) else 40000

    let calculateCredits performanceType audience = 
        Math.Max(audience - 30, 0) + if performanceType = "comedy" then audience / 5 else 0

    let createStatement (plays:Map<string, Play>, performance, formatLine) : Statement = 
        let performanceType = plays.[performance.PlayId].Type
        let amount = 
            match performanceType with
                | "comedy" -> calculateComedyPrice performance.Audience
                | "tragedy" -> calculateTragedyPrice performance.Audience
                | _ -> raise(System.Exception("unknown type: " + performanceType))

        let credits = calculateCredits performanceType performance.Audience
        let line = formatLine (plays.[performance.PlayId].Name) amount performance.Audience

        { Text= line; Amount = amount; Credits = credits }

    let printStatement (invoice, plays:Map<string, Play>) =
        invoice.Performances
            |> Seq.map(fun performance -> createStatement(plays, performance, formatLineToText))
            |> Seq.reduce(fun context line -> { Text = context.Text + line.Text; Amount = context.Amount + line.Amount; Credits = context.Credits + line.Credits })
            |> fun statement -> formatStatement statement invoice.Customer
   